File: completable-reactor-graph-viewer/src/main/kotlin/ru/fix/completable/reactor/graph/viewer/gl/GraphViewer.java
package ru.fix.completable.reactor.graph.viewer.gl;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import javafx.scene.Scene;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.val;
import ru.fix.completable.reactor.api.ReactorGraphModel;
import ru.fix.completable.reactor.api.gl.model.Source;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Created by Kamil Asfandiyarov.
 */
public class GraphViewer {
    ObjectMapper marshaller = new ObjectMapper();

    {
        marshaller.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    Scene scene;
    CopyOnWriteArrayList<ActionListener> actionListeners = new CopyOnWriteArrayList<>();
    GraphViewPane graphViewPane;

    public GraphViewer() {
        ActionListener actionListener = new ActionListener() {
            @Override
            public void goToSource(ReactorGraphModel.Source source) {
                for (val listener : actionListeners) {
                    listener.goToSource(source);
                }
            }

            @Override
            public void goToSubgraph(String subgraphPayloadClass) {
                for (val listener : actionListeners) {
                    listener.goToSubgraph(subgraphPayloadClass);
                }
            }

            @Override
            public void coordinatesChanged(List<CoordinateItem> coordinateItems) {
                for (val listener : actionListeners) {
                    listener.coordinatesChanged(coordinateItems);
                }
            }
        };

        graphViewPane = new GraphViewPane(actionListener, this::getShortcut);

        graphViewPane.setPrefSize(700.0, 600.0);

        scene = new Scene(graphViewPane);

        scene.getStylesheets().add(getClass().getResource("/css/styles.css").toExternalForm());

        //Shortcuts
        graphViewPane.setOnKeyReleased(keyEvent -> {
            shortcuts.forEach((shortcut, shortcutType) -> {
                if (shortcut.getPredicate().test(keyEvent)) {
                    switch (shortcutType) {
                        case GOTO_SERIALIZATION_POINT:
                            actionListener.goToSource(graphViewPane.getGraph().serializationPointSource);
                            break;
                        default:
                            throw new IllegalStateException();
                    }
                }

            });
        });
    }

    public void openGraph(ReactorGraphModel graph) {
        fixCoordinates(graph);
        graphViewPane.setGraph(graph);
    }

    int DEFAULT_POSITION = 100;

    /**
     * fix default coordinates on nodes in graph
     *
     * @param graph
     */
    public void fixCoordinates(ReactorGraphModel graph) {
        TreeNode treeNode = new TreeNode(graph.startPoint);
        for (val identity : graph.startPoint.processingItems) {
            recursiveBuldTree(graph, treeNode, identity);
        }
        recursiveFixCoordinates(treeNode,
                graph.startPoint.coordinates.x, graph.startPoint.coordinates.y, 200, 100);
    }


    public void openGraph(String graph) throws Exception {
        val graphModel = marshaller.readValue(graph, ReactorGraphModel.class);
        openGraph(graphModel);
    }

    public ReactorGraphModel getGraphModel() {
        return graphViewPane.getGraph();
    }

    public GraphViewer registerListener(ActionListener actionListener) {
        actionListeners.add(actionListener);
        return this;
    }

    Map<Shortcut, ShortcutType> shortcuts = new ConcurrentHashMap<>();

    public void setShortcut(ShortcutType shortcutType, Shortcut shortcut) {
        shortcuts.put(shortcut, shortcutType);
    }

    public Optional<Shortcut> getShortcut(ShortcutType shortcutType) {
        return shortcuts.entrySet().stream()
                .filter(entry -> entry.getValue() == shortcutType)
                .map(entry -> entry.getKey())
                .findFirst();
    }



    public interface ActionListener {

        /**
         * Viewer asks IDE to navigate to source code
         *
         * @param source source code location
         */
        void goToSource(Source source);

        /**
         * Viewer asks IDE to navigate to subgraph view
         * @param subgraphPayloadClass payload class name
         */
        void goToSubgraph(String subgraphPayloadClass);

        /**
         * Graph nodes coordinates changed
         *
         * @param coordinateItems new coordinates
         */
        void coordinatesChanged(List<CoordinateItem> coordinateItems);
    }

    public Scene getScene() {
        return scene;
    }
}

File: completable-reactor-graph-viewer/src/main/kotlin/ru/fix/completable/reactor/graph/viewer/gl/GraphViewer.kt
package ru.fix.completable.reactor.graph.viewer.gl

import javafx.scene.Scene
import javafx.scene.control.TabPane
import javafx.scene.layout.Pane
import ru.fix.completable.reactor.model.GraphModel
import ru.fix.completable.reactor.model.Source
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Created by Kamil Asfandiyarov.
 */
class GraphViewer {
    var scene: Scene
        private set

    private var actionListeners = CopyOnWriteArrayList<ActionListener>()

    private val viewPaneActionListener = object : ActionListener {
        override fun goToSource(source: Source) {
            for (listener in actionListeners) {
                listener.goToSource(source)
            }
        }

        override fun goToSubgraph(subgraphPayloadClass: String) {
            for (listener in actionListeners) {
                listener.goToSubgraph(subgraphPayloadClass)
            }
        }

        override fun coordinatesChanged(coordinateItems: List<CoordinateItem>) {
            for (listener in actionListeners) {
                listener.coordinatesChanged(coordinateItems)
            }
        }
    }

    private val shortcuts: MutableMap<Shortcut, ShortcutType> = ConcurrentHashMap()

    init {
        scene = Scene(Pane())
        scene.stylesheets.add(javaClass.getResource("/css/styles.css").toExternalForm())
    }

    fun openGraph(graphs: List<GraphModel>) {
        if (graphs.size == 1) {

            var graphViewPane = GraphViewPane(viewPaneActionListener, { this.getShortcut(it) })
            graphViewPane.setPrefSize(700.0, 600.0)

            scene.root = graphViewPane

            graphViewPane.openGraph(graphs[0])

            //Shortcuts
            addShortcutListener(graphViewPane)

        } else {
            val tabPane = TabPane()
            scene.root = tabPane

            for (graph in graphs) {
                var graphViewPane = GraphViewPane(viewPaneActionListener, { this.getShortcut(it) })
                graphViewPane.setPrefSize(700.0, 600.0)
                graphViewPane.openGraph(graph)

                addShortcutListener(graphViewPane)
            }
        }
    }

    private fun addShortcutListener(pane: GraphViewPane) {
        pane.setOnKeyReleased { keyEvent ->
            shortcuts.forEach { shortcut, shortcutType ->
                if (shortcut.getPredicate().test(keyEvent)) {
                    when (shortcutType) {
                        ShortcutType.GOTO_GRAPH -> pane.graphModel?.startPoint?.source?.let {
                            viewPaneActionListener.goToSource(it)
                        }
                    }
                }

            }
        }
    }


    fun registerListener(actionListener: ActionListener): GraphViewer {
        actionListeners.add(actionListener)
        return this
    }

    fun setShortcut(shortcutType: ShortcutType, shortcut: Shortcut) {
        shortcuts[shortcut] = shortcutType
    }

    fun getShortcut(shortcutType: ShortcutType): Shortcut? {
        for ((shortcut, type) in shortcuts) {
            if (type == shortcutType) {
                return shortcut
            }
        }
        return null
    }

    interface ActionListener {
        /**
         * Viewer asks IDE to navigate to source code
         *
         * @param source source code location
         */
        fun goToSource(source: Source)

        /**
         * Viewer asks IDE to navigate to subgraph view
         * @param subgraphPayloadType payload class name
         */
        fun goToSubgraph(subgraphPayloadType: String)

        /**
         * Graph nodes coordinates changed
         *
         * @param coordinateItems new coordinates
         */
        fun coordinatesChanged(coordinateItems: List<CoordinateItem>)
    }
}